pkg/api/authn_test.go

```diff
@@ -866,6 +867,326 @@ func TestAPIKeys(t *testing.T) {
+		defer cm.StopServer()
+
+		Convey("mTLS only authentication", func() {
+			// Test without client certificate - should fail
+			caCertPEM, err := os.ReadFile(caCertPath)
+			So(err, ShouldBeNil)
+
+			caCertPool := x509.NewCertPool()
+			caCertPool.AppendCertsFromPEM(caCertPEM)
+
+			client := resty.New()
+			client.SetTLSClientConfig(&tls.Config{RootCAs: caCertPool, MinVersion: tls.VersionTLS13})
+			_, err = client.R().Get(baseURL + "/v2/test-repo/tags/list")
+			So(err, ShouldNotBeNil)
+			So(err.Error(), ShouldContainSubstring, "tls: certificate required")
+
+			// Test with valid client certificate - should succeed
+			clientCert, err := tls.LoadX509KeyPair(clientCertPath, clientKeyPath)
+			So(err, ShouldBeNil)
+
+			client = resty.New()
+			client.SetTLSClientConfig(&tls.Config{
+				MinVersion:   tls.VersionTLS13,
+				Certificates: []tls.Certificate{clientCert},
+				RootCAs:      caCertPool,
+			})
+
+			resp, err := client.R().Get(baseURL + "/v2/test-repo/tags/list")
+			So(err, ShouldBeNil)
+			So(resp.StatusCode(), ShouldEqual, http.StatusNotFound) // 404 meaning we successfully passed auth
+
+			// Test with self-signed client certificate - should fail
+			selfSignedClientCert, err := tls.LoadX509KeyPair(selfSignedClientCertPath, selfSignedClientKeyPath)
+			So(err, ShouldBeNil)
+
+			client = resty.New()
+			client.SetTLSClientConfig(&tls.Config{
+				MinVersion:   tls.VersionTLS13,
+				Certificates: []tls.Certificate{selfSignedClientCert},
+				RootCAs:      caCertPool,
+			})
+
+			_, err = client.R().Get(baseURL + "/v2/test-selfsigned-repo/tags/list")
+			So(err, ShouldNotBeNil)
+			So(err.Error(), ShouldContainSubstring, "tls: certificate required")
+		})
+	})
+
+	Convey("Test mTLS with basic auth authentication", t, func() {
+		// Create temporary directory for certificates
+		tempDir := t.TempDir()
+
+		// Generate CA certificate
+		caCert, caKey, err := tlsutils.GenerateCACert()
+		So(err, ShouldBeNil)
+		caCertPath := path.Join(tempDir, "ca.crt")
+		err = os.WriteFile(caCertPath, caCert, 0o600)
+		So(err, ShouldBeNil)
+
+		// Generate server certificate
+		serverCertPath := path.Join(tempDir, "server.crt")
+		serverKeyPath := path.Join(tempDir, "server.key")
+		err = tlsutils.GenerateServerCertToFile("localhost", caCert, caKey, serverCertPath, serverKeyPath)
+		So(err, ShouldBeNil)
+
+		// Generate valid client certificate for "testuser" user
+		clientCertPath := path.Join(tempDir, "client.crt")
+		clientKeyPath := path.Join(tempDir, "client.key")
+		err = tlsutils.GenerateCertWithCNToFile("testuser", caCert, caKey, clientCertPath, clientKeyPath)
+		So(err, ShouldBeNil)
+
+		// Generate self-signed client cert for "testuser" user
+		selfSignedClientCertPath := path.Join(tempDir, "client-selfsigned.crt")
+		selfSignedClientKeyPath := path.Join(tempDir, "client-selfsigned.key")
+		err = tlsutils.GenerateSelfSignedCertWithCNToFile("testuser", selfSignedClientCertPath, selfSignedClientKeyPath)
+		So(err, ShouldBeNil)
```

Comment:
The code here that builds the certificates looks pretty similar to the code in the previous Convey. Perhaps having a small helper function might help to reduce some repetition.